icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,513 @@
+        U_ASSERT(utextPos >= prevUTextPos);
+
+        if (utextPos > prevUTextPos) {
+            if (utextPos != rangeStart
+                || (utextPos > 0
+                       && fClosePunctuationSet.contains(utext_char32At(inText, utextPos - 1)))) {
+                foundBreaks.push(utextPos, status);
+                correctedNumBreaks++;
+            }
+        }
+        else {
+            // Normalization expanded the input text, the dictionary found a boundary
+            // within the expansion, giving two boundaries with the same index in the
+            // original text. Ignore the second. See ticket #12918.
+            --numBreaks;
+        }
+        prevCPPos = cpPos;
+        prevUTextPos = utextPos;
+    }
+    (void)prevCPPos; // suppress compiler warnings about unused variable
+
+    UChar32 nextChar = utext_char32At(inText, rangeEnd);
+    if (!foundBreaks.isEmpty() && foundBreaks.peeki() == rangeEnd) {
+        // In phrase breaking, there has to be a breakpoint between Cj character and
+        // the number/open punctuation.
+        // E.g. る文字「そうだ、京都」->る▁文字▁「そうだ、▁京都」-> breakpoint between 字 and「
+        // E.g. 乗車率９０％程度だろうか -> 乗車▁率▁９０％▁程度だろうか -> breakpoint between 率 and ９
+        // E.g. しかもロゴがＵｎｉｃｏｄｅ！ -> しかも▁ロゴが▁Ｕｎｉｃｏｄｅ！-> breakpoint between が and Ｕ
+        if (!fDigitOrOpenPunctuationOrAlphabetSet.contains(nextChar)) {
+            foundBreaks.popi();
+            correctedNumBreaks--;
+        }
+    }
+
+    return correctedNumBreaks;
+}
+
+void CjMLBreakEngine::evaluateBreakpoint(const UVector &elementList, int32_t index,
+                                         int32_t &numBreaks, UVector32 &boundary,
+                                         UErrorCode &status) const {
+    UVector featureList(uprv_deleteUObject, uhash_compareUnicodeString, 42, status);
+    if (U_FAILURE(status)) {
+        return;
+    }
+
+    UChar32 arr[4] = {-1, -1, -1, -1};
+    int32_t length = 0;
+    const UChar32 w1 = ((Element *)elementList.elementAt(0))->ch;
+    const UChar32 w2 = ((Element *)elementList.elementAt(1))->ch;
+    const UChar32 w3 = ((Element *)elementList.elementAt(2))->ch;
+    const UChar32 w4 = ((Element *)elementList.elementAt(3))->ch;
+    const UChar32 w5 = ((Element *)elementList.elementAt(4))->ch;
+    const UChar32 w6 = ((Element *)elementList.elementAt(5))->ch;
+    const UnicodeString b1 = ((Element *)elementList.elementAt(0))->index;
+    const UnicodeString b2 = ((Element *)elementList.elementAt(1))->index;
+    const UnicodeString b3 = ((Element *)elementList.elementAt(2))->index;
+    const UnicodeString b4 = ((Element *)elementList.elementAt(3))->index;
+    const UnicodeString b5 = ((Element *)elementList.elementAt(4))->index;
+    const UnicodeString b6 = ((Element *)elementList.elementAt(5))->index;
+
+    if (w1 != INVALID) {
+        arr[0] = w1;
+        length = 1;
+        featureList.adoptElement(concatChar(u"UW1:", arr, length, status), status);
```

Comment:
You do not need to check status after these code right?
I am not sure why you need to have UChar32 arr[4] and  int32_t length = 0;
could you just change code to someting like
```
if (w1 != INVALID) {
        featureList.adoptElement(concatChar(u"UW1:", {w1}, 1, status), status);        
}
....
if (w2 != INVALID && w3 != INVALID) {
        featureList.adoptElement(concatChar(u"BW1:", {w2, w3}, 2, status), status);
}
...
if (w1 != INVALID && w2 != INVALID && w3 != INVALID) {
        featureList.adoptElement(concatChar(u"TW1:", {w1, w2, w3}, 3, status), status);
 }
```

Reply:
Arrays can only be initialized on definition. Afterwards, I could only initialize each entry manually.
Following your suggestion, there are compilation errors.
cjmlbe.cpp:253:44: error: invalid conversion from ‘UChar32’ {aka ‘int’} to ‘const UChar32*’ {aka ‘const int*’} [-fpermissive]
  253 |         featureList.adoptElement(concatChar(u"UW1:", {w1}, 1, status), status);


cjmlbe.cpp:271:44: error: cannot convert ‘<brace-enclosed initializer list>’ to ‘const UChar32*’ {aka ‘const int*’}
  271 |         featureList.adoptElement(concatChar(u"BW1:", {w2, w3}, 2, status), status);

Reply:
Per our discussion, keep the current implementation.